Design and Analysis of Algorithms – CSE2012  
Slot: L7 + L8  
Vishruth Devan – 20BCE2953  
This notebook can be found [here](https://github.com/vishruthdevan/vit-assignments/blob/main/DAA/DA3/classdemolab3.ipynb)


## Predicting Whether a Tumour is Benign or Malignant using K-Nearest Neighbours Algorithm {-}

In [1]:
import numpy as np
import pandas as pd # use for reading and processing the data
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier

In [2]:
bcData = pd.read_csv('wisc_bc_data.csv')

In [3]:
bcData.shape

(569, 32)

In [4]:
bcData.dtypes

id                     int64
diagnosis             object
radius_mean          float64
texture_mean         float64
perimeter_mean       float64
area_mean            float64
smoothness_mean      float64
compactness_mean     float64
concavity_mean       float64
points_mean          float64
symmetry_mean        float64
dimension_mean       float64
radius_se            float64
texture_se           float64
perimeter_se         float64
area_se              float64
smoothness_se        float64
compactness_se       float64
concavity_se         float64
points_se            float64
symmetry_se          float64
dimension_se         float64
radius_worst         float64
texture_worst        float64
perimeter_worst      float64
area_worst           float64
smoothness_worst     float64
compactness_worst    float64
concavity_worst      float64
points_worst         float64
symmetry_worst       float64
dimension_worst      float64
dtype: object

In [5]:
bcData.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0  87139402         B        12.32         12.39           78.85      464.1   
1   8910251         B        10.60         18.95           69.28      346.4   
2    905520         B        11.04         16.83           70.92      373.2   
3    868871         B        11.28         13.39           73.00      384.8   
4   9012568         B        15.19         13.21           97.65      711.8   

   smoothness_mean  compactness_mean  concavity_mean  points_mean  ...  \
0          0.10280           0.06981         0.03987      0.03700  ...   
1          0.09688           0.11470         0.06387      0.02642  ...   
2          0.10770           0.07804         0.03046      0.02480  ...   
3          0.11640           0.11360         0.04635      0.04796  ...   
4          0.07963           0.06934         0.03393      0.02657  ...   

   radius_worst  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0         13.50          15.64            86.97       549.1            0.1385   
1         11.88          22.94            78.28       424.8            0.1213   
2         12.41          26.44            79.93       471.4            0.1369   
3         11.92          15.77            76.53       434.0            0.1367   
4         16.20          15.73           104.50       819.1            0.1126   

   compactness_worst  concavity_worst  points_worst  symmetry_worst  \
0             0.1266          0.12420       0.09391          0.2827   
1             0.2515          0.19160       0.07926          0.2940   
2             0.1482          0.10670       0.07431          0.2998   
3             0.1822          0.08669       0.08611          0.2102   
4             0.1737          0.13620       0.08178          0.2487   

   dimension_worst  
0          0.06771  
1          0.07587  
2          0.07881  
3          0.06784  
4          0.06766  

[5 rows x 32 columns]

In [6]:
bcData.tail()

id diagnosis  radius_mean  texture_mean  perimeter_mean  \
564  911320502         B        13.17         18.22           84.28   
565     898677         B        10.26         14.71           66.20   
566     873885         M        15.28         22.41           98.92   
567     911201         B        14.53         13.98           93.86   
568    9012795         M        21.37         15.10          141.30   

     area_mean  smoothness_mean  compactness_mean  concavity_mean  \
564      537.3          0.07466           0.05994         0.04859   
565      321.6          0.09882           0.09159         0.03581   
566      710.6          0.09057           0.10520         0.05375   
567      644.2          0.10990           0.09242         0.06895   
568     1386.0          0.10010           0.15150         0.19320   

     points_mean  ...  radius_worst  texture_worst  perimeter_worst  \
564      0.02870  ...         14.90          23.89            95.10   
565      0.02037  ...         10.88          19.48            70.89   
566      0.03263  ...         17.80          28.03           113.80   
567      0.06495  ...         15.80          16.93           103.10   
568      0.12550  ...         22.69          21.84           152.10   

     area_worst  smoothness_worst  compactness_worst  concavity_worst  \
564       687.6            0.1282             0.1965          0.18760   
565       357.1            0.1360             0.1636          0.07162   
566       973.1            0.1301             0.3299          0.36300   
567       749.9            0.1347             0.1478          0.13730   
568      1535.0            0.1192             0.2840          0.40240   

     points_worst  symmetry_worst  dimension_worst  
564       0.10450          0.2235          0.06925  
565       0.04074          0.2434          0.08488  
566       0.12260          0.3175          0.09772  
567       0.10690          0.2606          0.07810  
568       0.19660          0.2730          0.08666  

[5 rows x 32 columns]

In [7]:
bcData['diagnosis'] = bcData.diagnosis.astype('category')

In [8]:
bcData.describe().transpose()

count          mean           std          min  \
id                 569.0  3.037183e+07  1.250206e+08  8670.000000   
radius_mean        569.0  1.412729e+01  3.524049e+00     6.981000   
texture_mean       569.0  1.928965e+01  4.301036e+00     9.710000   
perimeter_mean     569.0  9.196903e+01  2.429898e+01    43.790000   
area_mean          569.0  6.548891e+02  3.519141e+02   143.500000   
smoothness_mean    569.0  9.636028e-02  1.406413e-02     0.052630   
compactness_mean   569.0  1.043410e-01  5.281276e-02     0.019380   
concavity_mean     569.0  8.879932e-02  7.971981e-02     0.000000   
points_mean        569.0  4.891915e-02  3.880284e-02     0.000000   
symmetry_mean      569.0  1.811619e-01  2.741428e-02     0.106000   
dimension_mean     569.0  6.279761e-02  7.060363e-03     0.049960   
radius_se          569.0  4.051721e-01  2.773127e-01     0.111500   
texture_se         569.0  1.216853e+00  5.516484e-01     0.360200   
perimeter_se       569.0  2.866059e+00  2.021855e+00     0.757000   
area_se            569.0  4.033708e+01  4.549101e+01     6.802000   
smoothness_se      569.0  7.040979e-03  3.002518e-03     0.001713   
compactness_se     569.0  2.547814e-02  1.790818e-02     0.002252   
concavity_se       569.0  3.189372e-02  3.018606e-02     0.000000   
points_se          569.0  1.179614e-02  6.170285e-03     0.000000   
symmetry_se        569.0  2.054230e-02  8.266372e-03     0.007882   
dimension_se       569.0  3.794904e-03  2.646071e-03     0.000895   
radius_worst       569.0  1.626919e+01  4.833242e+00     7.930000   
texture_worst      569.0  2.567722e+01  6.146258e+00    12.020000   
perimeter_worst    569.0  1.072612e+02  3.360254e+01    50.410000   
area_worst         569.0  8.805831e+02  5.693570e+02   185.200000   
smoothness_worst   569.0  1.323686e-01  2.283243e-02     0.071170   
compactness_worst  569.0  2.542650e-01  1.573365e-01     0.027290   
concavity_worst    569.0  2.721885e-01  2.086243e-01     0.000000   
points_worst       569.0  1.146062e-01  6.573234e-02     0.000000   
symmetry_worst     569.0  2.900756e-01  6.186747e-02     0.156500   
dimension_worst    569.0  8.394582e-02  1.806127e-02     0.055040   

                             25%            50%           75%           max  
id                 869218.000000  906024.000000  8.813129e+06  9.113205e+08  
radius_mean            11.700000      13.370000  1.578000e+01  2.811000e+01  
texture_mean           16.170000      18.840000  2.180000e+01  3.928000e+01  
perimeter_mean         75.170000      86.240000  1.041000e+02  1.885000e+02  
area_mean             420.300000     551.100000  7.827000e+02  2.501000e+03  
smoothness_mean         0.086370       0.095870  1.053000e-01  1.634000e-01  
compactness_mean        0.064920       0.092630  1.304000e-01  3.454000e-01  
concavity_mean          0.029560       0.061540  1.307000e-01  4.268000e-01  
points_mean             0.020310       0.033500  7.400000e-02  2.012000e-01  
symmetry_mean           0.161900       0.179200  1.957000e-01  3.040000e-01  
dimension_mean          0.057700       0.061540  6.612000e-02  9.744000e-02  
radius_se               0.232400       0.324200  4.789000e-01  2.873000e+00  
texture_se              0.833900       1.108000  1.474000e+00  4.885000e+00  
perimeter_se            1.606000       2.287000  3.357000e+00  2.198000e+01  
area_se                17.850000      24.530000  4.519000e+01  5.422000e+02  
smoothness_se           0.005169       0.006380  8.146000e-03  3.113000e-02  
compactness_se          0.013080       0.020450  3.245000e-02  1.354000e-01  
concavity_se            0.015090       0.025890  4.205000e-02  3.960000e-01  
points_se               0.007638       0.010930  1.471000e-02  5.279000e-02  
symmetry_se             0.015160       0.018730  2.348000e-02  7.895000e-02  
dimension_se            0.002248       0.003187  4.558000e-03  2.984000e-02  
radius_worst           13.010000      14.970000  1.879000e+01  3.604000e+01  
texture_worst   

In [9]:
bcData.groupby(['diagnosis']).count()

id  radius_mean  texture_mean  perimeter_mean  area_mean  \
diagnosis                                                              
B          357          357           357             357        357   
M          212          212           212             212        212   

           smoothness_mean  compactness_mean  concavity_mean  points_mean  \
diagnosis                                                                   
B                      357               357             357          357   
M                      212               212             212          212   

           symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
diagnosis                 ...                                                 
B                    357  ...           357            357              357   
M                    212  ...           212            212              212   

           area_worst  smoothness_worst  compactness_worst  concavity_worst  \
diagnosis                                                                     
B                 357               357                357              357   
M                 212               212                212              212   

           points_worst  symmetry_worst  dimension_worst  
diagnosis                                                 
B                   357             357              357  
M                   212             212              212  

[2 rows x 31 columns]

In [10]:
bcData = bcData.drop(labels=['id'], axis = 1)

In [11]:
bcData.head()

diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0         B        12.32         12.39           78.85      464.1   
1         B        10.60         18.95           69.28      346.4   
2         B        11.04         16.83           70.92      373.2   
3         B        11.28         13.39           73.00      384.8   
4         B        15.19         13.21           97.65      711.8   

   smoothness_mean  compactness_mean  concavity_mean  points_mean  \
0          0.10280           0.06981         0.03987      0.03700   
1          0.09688           0.11470         0.06387      0.02642   
2          0.10770           0.07804         0.03046      0.02480   
3          0.11640           0.11360         0.04635      0.04796   
4          0.07963           0.06934         0.03393      0.02657   

   symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0         0.1959  ...         13.50          15.64            86.97   
1         0.1922  ...         11.88          22.94            78.28   
2         0.1714  ...         12.41          26.44            79.93   
3         0.1771  ...         11.92          15.77            76.53   
4         0.1721  ...         16.20          15.73           104.50   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0       549.1            0.1385             0.1266          0.12420   
1       424.8            0.1213             0.2515          0.19160   
2       471.4            0.1369             0.1482          0.10670   
3       434.0            0.1367             0.1822          0.08669   
4       819.1            0.1126             0.1737          0.13620   

   points_worst  symmetry_worst  dimension_worst  
0       0.09391          0.2827          0.06771  
1       0.07926          0.2940          0.07587  
2       0.07431          0.2998          0.07881  
3       0.08611          0.2102          0.06784  
4       0.08178          0.2487          0.06766  

[5 rows x 31 columns]

In [12]:
X = bcData.drop(labels=['diagnosis'], axis = 1)
y = bcData['diagnosis']
print(X.head())
print(y.head())

   radius_mean  texture_mean  perimeter_mean  area_mean  smoothness_mean  \
0        12.32         12.39           78.85      464.1          0.10280   
1        10.60         18.95           69.28      346.4          0.09688   
2        11.04         16.83           70.92      373.2          0.10770   
3        11.28         13.39           73.00      384.8          0.11640   
4        15.19         13.21           97.65      711.8          0.07963   

   compactness_mean  concavity_mean  points_mean  symmetry_mean  \
0           0.06981         0.03987      0.03700         0.1959   
1           0.11470         0.06387      0.02642         0.1922   
2           0.07804         0.03046      0.02480         0.1714   
3           0.11360         0.04635      0.04796         0.1771   
4           0.06934         0.03393      0.02657         0.1721   

   dimension_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0         0.05955  ...         13.50          15.64            86.97

In [13]:
zscorenormalization = lambda x : (x-x.mean())/x.std()
XNormalized = X.apply(zscorenormalization)
XNormalized.describe()

radius_mean  texture_mean  perimeter_mean     area_mean  \
count  5.690000e+02  5.690000e+02    5.690000e+02  5.690000e+02   
mean  -1.632750e-15  2.513709e-15    3.370863e-15  1.311195e-16   
std    1.000000e+00  1.000000e+00    1.000000e+00  1.000000e+00   
min   -2.027864e+00 -2.227289e+00   -1.982759e+00 -1.453164e+00   
25%   -6.887793e-01 -7.253249e-01   -6.913472e-01 -6.666089e-01   
50%   -2.148925e-01 -1.045442e-01   -2.357726e-01 -2.949274e-01   
75%    4.689800e-01  5.836621e-01    4.992377e-01  3.631877e-01   
max    3.967796e+00  4.647799e+00    3.972634e+00  5.245913e+00   

       smoothness_mean  compactness_mean  concavity_mean   points_mean  \
count     5.690000e+02      5.690000e+02    5.690000e+02  5.690000e+02   
mean     -1.803771e-15     -1.422803e-15    2.200934e-16  1.509825e-15   
std       1.000000e+00      1.000000e+00    1.000000e+00  1.000000e+00   
min      -3.109349e+00     -1.608721e+00   -1.113893e+00 -1.260710e+00   
25%      -7.103378e-01     -7.464292e-01   -7.430941e-01 -7.372951e-01   
50%      -3.486040e-02     -2.217454e-01   -3.419391e-01 -3.973715e-01   
75%       6.356397e-01      4.934227e-01    5.255994e-01  6.463664e-01   
max       4.766717e+00      4.564409e+00    4.239858e+00  3.924477e+00   

       symmetry_mean  dimension_mean  ...  radius_worst  texture_worst  \
count   5.690000e+02    5.690000e+02  ...  5.690000e+02   5.690000e+02   
mean   -4.898249e-15    2.452686e-15  ... -7.765708e-17  -8.780323e-18   
std     1.000000e+00    1.000000e+00  ...  1.000000e+00   1.000000e+00   
min    -2.741705e+00   -1.818265e+00  ... -1.725382e+00  -2.222039e+00   
25%    -7.026215e-01   -7.220040e-01  ... -6.743279e-01  -7.479711e-01   
50%    -7.156354e-02   -1.781226e-01  ... -2.688030e-01  -4.347738e-02   
75%     5.303125e-01    4.705693e-01  ...  5.215568e-01   6.577623e-01   
max     4.480808e+00    4.906602e+00  ...  4.090590e+00   3.882489e+00   

       perimeter_worst    area_worst  smoothness_worst  compactness_worst  \
count     5.690000e+02  5.690000e+02      5.690000e+02       5.690000e+02   
mean      4.581377e-16 -1.592555e-15     -1.448948e-15       1.767235e-15   
std       1.000000e+00  1.000000e+00      1.000000e+00       1.000000e+00   
min      -1.691872e+00 -1.221348e+00     -2.680337e+00      -1.442609e+00   
25%      -6.889721e-01 -6.415713e-01     -6.906227e-01      -6.804845e-01   
50%      -2.857288e-01 -3.408813e-01     -4.680159e-02      -2.692639e-01   
75%       5.398040e-01  3.572747e-01      5.970195e-01       5.391944e-01   
max       4.283568e+00  5.924959e+00      3.951897e+00       5.108382e+00   

       concavity_worst  points_worst  symmetry_worst  dimension_worst  
count     5.690000e+02  5.690000e+02    5.690000e+02     5.690000e+02  
mean     -1.880940e-16  2.200934e-16   -6.365734e-16    -1.314219e-15  
std       1.000000e+00  1.000000e+00    1.000000e+00     1.000000e+00  
min      -1.304683e+00 -1.743529e+00   -2.159060e+00    -1.600431e+00  
25%      -7.558491e-01 -7.557349e-01   -6.412994e-01    -6.913035e-01  
50%      -2.180402e-01 -2.232725e-01   -1.272975e-01    -2.162538e-01  
75%       5.306742e-01  7.118836e-01    4.497425e-01     4.503661e-01  
max       4.696536e+00  2.683516e+00    6.040726e+00     6.840837e+00  

[8 rows x 30 columns]

In [14]:
X_train, X_test, y_train, y_test = train_test_split(XNormalized, y, test_size=0.3, random_state=1)

In [15]:
X_train

radius_mean  texture_mean  perimeter_mean  area_mean  smoothness_mean  \
249    -0.603650      2.078651       -0.625501  -0.603525        -0.852544   
58      2.872465      0.211659        3.054900   3.143127         3.437093   
476    -0.589462      0.797564       -0.544016  -0.588465        -1.920509   
529    -1.265389     -0.276596       -1.272030  -1.047384        -0.942133   
422    -0.844850     -1.443757       -0.868309  -0.775726         0.083881   
..           ...           ...             ...        ...              ...   
129     0.040496      0.758039        0.074117  -0.071293         0.528985   
144     0.128463      0.520887        0.223917  -0.028669         0.642750   
72     -1.443025     -0.090594       -1.312772  -1.166163         0.236042   
235    -0.694455     -0.725325       -0.678178  -0.666041         1.168911   
37     -0.632026     -1.078728       -0.570355  -0.631089         1.339558   

     compactness_mean  concavity_mean  points_mean  symmetry_mean  \
249         -0.754382       -0.605236    -0.759458      -0.717212   
58           3.452935        4.239858     3.924477       3.076431   
476          0.056028       -0.117528    -0.493241      -2.220079   
529         -0.947896       -0.928368    -1.112036      -0.480110   
422         -1.007540       -0.865272    -0.800435       0.067050   
..                ...             ...          ...            ...   
129          0.209400        0.721285     0.321390       0.417233   
144          1.561347        0.673618     1.002784       1.606394   
72           1.758269        1.363534     0.004145       1.890917   
235         -0.221745       -0.577138    -0.453553       0.150948   
37           0.478275       -0.648513    -0.486798       0.668926   

     dimension_mean  ...  radius_worst  texture_worst  perimeter_worst  \
249       -0.248940  ...     -0.657776       1.985399        -0.660403   
58         0.845904  ...      2.017447      -0.274512         2.191465   
476        0.537138  ...     -0.765364       0.493435        -0.592253   
529        0.187298  ...     -1.059163      -0.172011        -1.075848   
422       -0.247524  ...     -0.885366      -1.525680        -0.922883   
..              ...  ...           ...            ...              ...   
129       -0.421736  ...     -0.202595       1.398050        -0.088125   
144        0.912473  ...      0.279483       1.225588         0.450525   
72         3.488828  ...     -1.284684      -0.370506        -1.149949   
235        0.171718  ...     -0.610189      -0.664994        -0.615763   
37         2.208440  ...     -0.645362      -1.491513        -0.624989   

     area_worst  smoothness_worst  compactness_worst  concavity_worst  \
249   -0.626818         -0.506674          -0.436422        -0.436615   
58     2.094322          1.630637           1.081345         1.476873   
476   -0.688818         -1.943665           0.426697         0.091128   
529   -0.878857          0.281679          -0.818405        -0.997480   
422   -0.772421          0.075831          -1.045880        -0.963591   
..          ...               ...                ...              ...   
129   -0.267465          0.246641           0.120982         0.724803   
144    0.028658          0.881702           2.606102         1.350329   
72    -1.024811         -0.449737           0.766097         0.904552   
235   -0.580977          0.886082          -0.677307        -0.590480   
37    -0.651583          0.439349          -0.016303        -0.853153   

     points_worst  symmetry_worst  dimension_worst  
249     -0.660957       -0.733432        -0.111056  
58       1.676401        0.519246        -0.213485  
476     -0.082246       -1.147220         0.528434  
529     -1.363198        0.252547        -0.287677  
422     -0.905889       -0.067492        -0.898376  
..            ...             ...              ...  
129      0.328511       -0.327726        -0.024130  
144      2.365560        2.203491         2.41146

In [16]:
X_test

radius_mean  texture_mean  perimeter_mean  area_mean  smoothness_mean  \
421     0.276020      0.634813        0.217744   0.164560        -0.412417   
47      3.292437     -0.425397        3.384132   3.850686         1.318227   
292    -0.101387      0.697588       -0.055107  -0.187515         1.680852   
186     0.829361     -0.048744        0.881970   0.682300         1.261345   
414    -0.677429     -1.225205       -0.729620  -0.646718        -1.296225   
..           ...           ...             ...        ...              ...   
6      -0.742695      1.078892       -0.718097  -0.714348        -0.266656   
487     2.310612      0.088433        2.503437   2.427043         2.576748   
11     -0.532709     -0.313796       -0.563770  -0.552945        -0.698250   
268    -0.532709      0.732463       -0.567474  -0.535327        -1.026746   
419    -0.660403     -0.629999       -0.534139  -0.633362         0.905831   

     compactness_mean  concavity_mean  points_mean  symmetry_mean  \
421         -0.634903       -0.454960    -0.401495      -0.713565   
47           2.498620        3.110904     3.669341       0.526665   
292          0.423364        0.623442     0.421125       0.081641   
186          1.000876        1.281999     1.548362       1.165018   
414         -1.153906       -0.832281    -0.548649       1.055586   
..                ...             ...          ...            ...   
6           -0.042433        0.280993    -0.202798      -1.545248   
487          3.265480        4.234841     3.437399       2.715305   
11          -0.711021       -0.626561    -0.659981       0.577733   
268         -0.991067       -0.898890    -0.935734      -0.976202   
419          1.243242        0.967146     0.650232       3.160329   

     dimension_mean  ...  radius_worst  texture_worst  perimeter_worst  \
421       -0.843811  ...      0.378796       0.978283         0.310655   
47        -0.224862  ...      3.488510      -0.341220         3.631832   
292        0.845904  ...      0.024996       1.355748         0.129121   
186        0.064075  ...      0.687077      -0.128082         0.780857   
414       -0.236193  ...     -0.612258      -1.206787        -0.671414   
..              ...  ...           ...            ...              ...   
6          0.411082  ...     -0.783985       1.868255        -0.743432   
487        1.075354  ...      2.509870       0.379544         2.962240   
11        -0.073312  ...     -0.554326      -0.074391        -0.614871   
268       -0.430234  ...     -0.434323       1.027093        -0.432146   
419        1.545585  ...     -0.523291       0.114342        -0.456252   

     area_worst  smoothness_worst  compactness_worst  concavity_worst  \
421    0.262431          0.286058          -0.308670        -0.004738   
47     4.133464          0.903601           2.157382         1.787479   
292   -0.129415          1.810206           0.368223         0.521088   
186    0.541693          1.661295           0.884315         1.100598   
414   -0.609254         -1.259988          -1.075816        -0.981854   
..          ...               ...                ...              ...   
6     -0.713758         -0.112498          -0.016303         0.435287   
487    2.598399          1.652536           2.831098         3.300726   
11    -0.555685         -0.467256          -0.479641        -0.373343   
268   -0.452586         -0.296447          -0.469472        -0.715106   
419   -0.507385          0.268539           0.984736         0.854222   

     points_worst  symmetry_worst  dimension_worst  
421      0.584093       -0.364902        -0.288785  
47       2.449232        1.275702         0.232773  
292      0.801642        0.289723         0.519575  
186      2.125191        0.336597         0.368976  
414     -0.482353        0.323667        -0.944885  
..            ...             ...              ...  
6       -0.274997       -1.274912         0.186819  
487      2.683516        1.865672         0.77204

In [17]:
KNN = KNeighborsClassifier(n_neighbors = 5, weights='distance')

In [18]:
KNN.fit(X_train, y_train)

KNeighborsClassifier(weights='distance')

In [19]:
predicted_labels = KNN.predict(X_test)
print(predicted_labels)

['M' 'M' 'M' 'M' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'M' 'B' 'B'
 'B' 'B' 'B' 'B' 'M' 'M' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'M' 'B' 'B' 'B' 'B'
 'M' 'B' 'M' 'M' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'M' 'B' 'B' 'B'
 'B' 'M' 'M' 'M' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'M' 'B' 'M' 'B' 'B' 'B' 'B'
 'M' 'B' 'M' 'B' 'M' 'M' 'M' 'B' 'B' 'B' 'B' 'B' 'M' 'B' 'M' 'B' 'B' 'M'
 'M' 'B' 'M' 'B' 'B' 'B' 'M' 'B' 'M' 'B' 'B' 'M' 'M' 'B' 'B' 'M' 'B' 'M'
 'B' 'M' 'B' 'B' 'B' 'M' 'M' 'B' 'B' 'M' 'B' 'B' 'B' 'M' 'B' 'B' 'B' 'M'
 'M' 'M' 'B' 'B' 'B' 'B' 'M' 'M' 'B' 'B' 'B' 'M' 'B' 'B' 'M' 'M' 'M' 'B'
 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'B' 'M' 'B' 'M' 'M' 'B' 'B' 'B' 'B' 'M'
 'B' 'M' 'M' 'B' 'B' 'M' 'B' 'B' 'M']


In [20]:
KNN.score(X_test, y_test)

0.9649122807017544








## K-Nearest Neighbours Algorithm and Implementation {-}

In [21]:
# custom knn prediction algorithm

import math


def knn(points, p, k=5):
    distances = [] 
    # list that stores the euclidean distances of all the train data points from the given point p
    
    frequencies = {} 
    # dictionary that stores the number of points of each category neighbouring the given point p 
    
    for category in points:
        frequencies[category] = 0 # initially the number of neighbouring points is 0 
        # calculating the euclidean distance for all features(dimensions) of each point
        for point in points[category]:
            ed = 0  
            for i, j in zip(point, p):
                ed += (i-j)**2
            ed = math.sqrt(ed)
            distances.append((ed, category))
            
    distances = sorted(distances)[:k]
            
    # number of neighbouring points updated for each category
    for i in distances:
        frequencies[i[1]] += 1
    
    # finding the most frequent category in the k neighbouring points
    max = list(frequencies.keys())[0]
    for i in frequencies:
        if frequencies[i] > frequencies[max]:
            max = i
    
    # returning the most frequent category neighbouring the given point
    return max

In [22]:
# formatting the data to send as arguments to the knn() function

#converting dataframe values to a list
X_train_list = X_train.values.tolist()
y_train_list = y_train.values.tolist()


divided_by_category = {x:[] for x in y_train.unique()}
# divided_by_category is a dictionary of the form:
# {'B': [list of points with category 'B'], 
#  'M': [list of points with category 'M']}


for i, j in zip(X_train_list, y_train_list):
    # appending the training points to one of the categories in the dictionary
    divided_by_category[j].append(i)

In [23]:
# using the knn() function to predict the categories

prediction = []
# prediction is a list that stores the category predicted by the knn() function from the given test data

for i in X_test.values.tolist():
    # calling knn() and appending the predicted category for each point in X_test
    prediction.append(knn(divided_by_category, i, 5))

print(prediction)

['M', 'M', 'M', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'M', 'B', 'M', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'M', 'M', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'B', 'M', 'B', 'B', 'B', 'B', 'M', 'B', 'M', 'B', 'M', 'M', 'M', 'B', 'B', 'B', 'B', 'B', 'M', 'B', 'M', 'B', 'B', 'M', 'M', 'B', 'M', 'B', 'B', 'B', 'M', 'B', 'M', 'B', 'B', 'M', 'M', 'B', 'B', 'M', 'B', 'M', 'B', 'M', 'B', 'B', 'B', 'M', 'M', 'B', 'B', 'M', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'M', 'M', 'M', 'B', 'B', 'B', 'B', 'M', 'M', 'B', 'B', 'B', 'M', 'B', 'B', 'M', 'M', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'B', 'M', 'M', 'B', 'B', 'B', 'B', 'M', 'B', 'M', 'M', 'B', 'B', 'M', 'B', 'B', 'M']


In [24]:
# checking the accuracy of the knn algorithm

count = 0
for i, j in zip(prediction, y_test):
    # if the predicted category matches the actual category from y_test, count is increased 
    if i==j:
        count += 1

print(count/len(y_test.values))
# dividing the number of categories predicted correctly by the total number of predictions, we get the accuracy

0.9649122807017544


The accuracy of predictions by our own implementation of the KNN algorithm is 0.9649122807017544.  